src/main/scala/akka/contrib/process/BlockingProcess.scala

```diff
@@ -97,77 +97,55 @@ class BlockingProcess(
+    val destroyer =
+      context.watch(context.actorOf(ProcessDestroyer.props(process, receiver), "process-destroyer"))
 
-  private val stdout =
-    context.watch(context.actorOf(InputStreamPublisher.props(process.getInputStream, stdioTimeout), "stdout"))
+    val stdin =
+      context.actorOf(OutputStreamSubscriber.props(process.getOutputStream), "stdin")
 
-  private val stderr =
-    context.watch(context.actorOf(InputStreamPublisher.props(process.getErrorStream, stdioTimeout), "stderr"))
+    val stdout =
+      context.watch(context.actorOf(InputStreamPublisher.props(process.getInputStream, stdioTimeout), "stdout"))
 
-  private var nrOfPublishers = 2
+    val stderr =
+      context.watch(context.actorOf(InputStreamPublisher.props(process.getErrorStream, stdioTimeout), "stderr"))
 
-  override def preStart(): Unit =
     receiver ! Started(ActorSubscriber(stdin), ActorPublisher(stdout), ActorPublisher(stderr))
+  }
 
   override def receive = {
     case Destroy =>
       log.debug("Received request to destroy the process.")
-      process.destroy()
-
-    case Terminated(publisher @ (`stdout` | `stderr`)) =>
-      log.debug("Publisher {} has terminated", publisher.path.name)
-      nrOfPublishers -= 1
-      if (nrOfPublishers == 0) {
-        val exitValue =
-          blocking {
-            process.waitFor()
-            process.exitValue()
-          }
-        receiver ! Exited(exitValue)
-        context.stop(self)
-      }
-  }
-
-  private def startProcess(): JavaProcess = {
-    import JavaConverters._
-    val pb = new JavaProcessBuilder(prepareCommand(command).asJava)
-    pb.environment().putAll(environment.asJava)
-    pb.directory(directory)
-    pb.start()
+      context.stop(self)
+    case Terminated(ref) =>
+      log.debug("Child {} was unexpectedly stopped, shutting down process", ref.path)
+      context.stop(self)
   }
-
-  private def prepareCommand(args: immutable.Seq[String]): immutable.Seq[String] =
-    if (Helpers.isWindows)
-      args map winQuote
-    else
-      args
 }
 
-private object ShutdownHook {
-  def props(op: () => Unit): Props =
-    Props(new ShutdownHook(op))
+private object ProcessDestroyer {
+  def props(process: JavaProcess, exitValueReceiver: ActorRef): Props =
+    Props(new ProcessDestroyer(process, exitValueReceiver))
 }
 
-/*
- * Responsible for performing a side-effecting operation when the actor sys is required to shutdown.
- */
-private class ShutdownHook(op: () => Unit) extends Actor {
-
+private class ProcessDestroyer(process: JavaProcess, exitValueReceiver: ActorRef) extends Actor {
   override def receive =
     Actor.emptyBehavior
 
-  override def postStop(): Unit =
-    try
-      op()
-    catch {
-      case NonFatal(_) =>
+  override def postStop(): Unit = {
+    val exitValue = blocking {
+      process.destroy()
+      process.waitFor()
     }
+    exitValueReceiver ! BlockingProcess.Exited(exitValue)
```

Comment:
@huntc Given the less-than-desired quality of the documentation of Process—no we can't be 100% certain, but if the Process terminated prior to destroy being executed, what exitValue would it invent, and why would it discard the (already known) 0? We -are- executing testcases that look at the exitValue so we should be able to spot if there's something fishy going on.